src/type_object.rs

```diff
@@ -65,27 +70,19 @@ pub unsafe trait PyTypeInfo: Sized {
     }
 }
 
-/// Python object types that have a corresponding type object.
+/// Legacy trait which previously held the `type_object` method now found on `PyTypeInfo`.
 ///
 /// # Safety
 ///
-/// This trait is marked unsafe because not fulfilling the contract for type_object
-/// leads to UB.
-///
-/// See also [PyTypeInfo::type_object_raw](trait.PyTypeInfo.html#tymethod.type_object_raw).
-pub unsafe trait PyTypeObject {
-    /// Returns the safe abstraction over the type object.
-    fn type_object(py: Python<'_>) -> &PyType;
-}
-
-unsafe impl<T> PyTypeObject for T
-where
-    T: PyTypeInfo,
-{
-    fn type_object(py: Python<'_>) -> &PyType {
-        unsafe { py.from_borrowed_ptr(Self::type_object_raw(py) as _) }
-    }
-}
+/// This trait used to have stringent safety requirements, but they are now irrelevant as it is deprecated.
+#[deprecated(
+    since = "0.17.0",
+    note = "PyTypeObject::type_object was moved to PyTypeInfo::type_object"
+)]
+pub unsafe trait PyTypeObject: PyTypeInfo {}
```

Comment:
Since the trait is changed in an incompatible way, is there actually a use where keep it around but deprecated instead of removing it outright would help? As written above, having trait bounds to access `type_object` needs to be updated after this. If there is any explicit implementor, it also needs changing after this. Which kind of usage would survive unchanged with the deprecated trait?

Reply:
I think now I possibly understand it: Existing trait bounds used to access `type_object` should continue to work due to having `PyTypeInfo` as a super trait? Albeit with a deprecation warning? Implementors would still be affected, but there should not be any useful manual ones due to conflicts with the existing blanket impl?

Reply:
That was the idea. It's not perfect, but at least it can serve as a waypoint for users to guide them towards how to migrate.

Pushed a test demonstrating `PyTypeObject` still working.